### DISCUSSION TITLE
Audio and Video streaming from file

### DISCUSSION THREAD

**Discussion by anil88** [devopvoid/webrtc-java#90] [Q&A]
Hello,

is it possible to stream an audio/video coming from a file?

I plan to write a test where one WebRTC client streams a predefined audio/video and another WebRTC client checks the received audio/video stream.

Br Anil

**Comment by devopvoid** [comment]
Hi,
this is possible, but I did not write any helper classes for this purpose. With the current state of the implementation you can achieve that by providing custom audio and video sources. But you have to do it by yourself:
1. Reading the file.
2. Demuxing / Decoding the media file into separate audio and video streams.
3. Provide the data from each stream with a custom source for each stream.
4. Optionally do the syncing.

Maybe the WebRTC sources do have an example for it. I didn't have time to tackle this issue yet.

**Reply by anil88** [reply]
Hi,

thank you for the fast response.

How can I create a custom source for the stream? Do you mean I have to create a custom AudioTrackSource for audio? If yes how?
```
 AudioTrackSource audioSource = factory.createAudioSource(audioOptions);
 AudioTrack audioTrack = factory.createAudioTrack("audioTrack", audioSource);
```

Br Anil

**Reply by ranaparti** [reply]
Hi @anil88,

I want to stream audio from a file in JAVA framework and tried this library. Later, I found that there is no way to set a custom audio source and noticed your question as well here.

Did you implement the custom source for the stream? How did you stream the audio through WebRTC?

Thanks,
Ramesh

**Reply by stserakhau** [reply]
At the moment I found only one way how to solve this.
1. Via ffmpeg (bytedeco) read frames & split to video and audio streams
2. Send media data via datachannel
3. Play video via jmuxer, play audio via basic audio web api

**Reply by devopvoid** [reply]
Update: This feature is work in progress. It will use FFmpeg for media decoding. FFmpeg itself will not be bundled with the release and can be optionally linked in the app, which enables this feature.

**Reply by kinsleykajiva** [reply]
any update on this ?

**Reply by florensie** [reply]
With a custom audio source it's just not possible to feed in data myself right now?
